datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1337,12 +1339,33 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
         _adapter.updateAllAssignments(newAssignmentsByInstance);
       }
 
+      Set<String> failedStreams = Collections.emptySet();
+      // Poll the zookeeper to ensure that hosts claimed assignment tokens for stopping streams
+      if (_config.getEnableAssignmentTokens() &&
+          !PollUtils.poll(() -> _adapter.getNumUnclaimedTokensForDatastreams(stoppingDatastreamGroups) == 0,
+            STOP_PROPAGATION_RETRY_MS, _config.getStopPropagationTimeout())) {
+        Map<String, List<AssignmentToken>> unclaimedTokens =
+            _adapter.getUnclaimedAssignmentTokensForDatastreams(stoppingDatastreamGroups);
+        failedStreams = unclaimedTokens.keySet();
+        Set<String> hosts = unclaimedTokens.values().stream().flatMap(List::stream).
+            map(AssignmentToken::getIssuedFor).collect(Collectors.toSet());
+        _log.error("Stop failed to propagate within {}ms for streams: {}. The following hosts failed to claim their token(s): {}",
+            _config.getStopPropagationTimeout(), failedStreams, hosts);
+        revokeUnclaimedAssignmentTokens(unclaimedTokens, stoppingDatastreamGroups);
+        _metrics.updateMeter(CoordinatorMetrics.Meter.NUM_FAILED_STOPS, failedStreams.size());
+      }
+
+      // TODO Explore introduction of UNKNOWN/WARN state for streams that failed to stop
+
       for (DatastreamGroup datastreamGroup : stoppingDatastreamGroups) {
         for (Datastream datastream : datastreamGroup.getDatastreams()) {
-          datastream.setStatus(DatastreamStatus.STOPPED);
-          if (!_adapter.updateDatastream(datastream)) {
-            _log.warn("Failed to update datastream: {} to stopped state", datastream.getName());
-            succeeded = false;
+          // Only streams that were confirmed to have stopped successfully will be transitioned to STOPPED state
+          if (!failedStreams.contains(datastream.getName())) {
+            datastream.setStatus(DatastreamStatus.STOPPED);
+            if (!_adapter.updateDatastream(datastream)) {
+              _log.warn("Failed to update datastream: {} to stopped state", datastream.getName());
+              succeeded = false;
```

Comment:
@shrinandthakkar Can you double check that the recent issue that you saw with the previous assignment getting replaced by new assignment and some cleanup is not completed, will not reoccur in this scenario?